src/System.Collections.Immutable/src/System/Collections/Immutable/ImmutableSortedSet`1.cs

```diff
@@ -1102,10 +1102,39 @@ private ImmutableSortedSet<T> LeafToRootRefill(IEnumerable<T> addedItems)
             // and can index into that sequence like a list, so the limited
             // garbage produced is a temporary mutable data structure we use
             // as a reference when creating the immutable one.
-            // The (mutable) SortedSet<T> is much faster at constructing its collection 
-            // when passed a sequence into its constructor than into its Union method.
-            var sortedSet = new SortedSet<T>(this.Concat(addedItems), this.KeyComparer);
-            Node root = Node.NodeTreeFromSortedSet(sortedSet);
+
+            // Produce the initial list containing all elements, including any duplicates.
+            List<T> list;
+            if (this.IsEmpty)
+            {
+                list = new List<T>(addedItems);
```

Comment:
Tip: I think we have a TryGetCount extension method available we use to sniff the length of an `IEnumerable<T>` if possible in order to avoid work such as your `new List<T>` here where we can. You might consider using that here.